### ISSUE TITLE
AppDomain.SetPrincipalPolicy(PrincipalPolicy.WindowsPrincipal) works only once

### ISSUE THREAD

**Issue by marcselis** [dotnet/runtime#31717]
Setting the PrincipalPolicy on the current AppDomain to WindowsPrincipal works only for the first thread being started.  Any subsequent thread has Thread.CurrentPrincipal evaluated to NULL.

See the following example:

```
    internal static class Program 
    {
        private static void Main()
        {
            // Set the principal policy to WindowsPrincipal.
            AppDomain.CurrentDomain.SetPrincipalPolicy(PrincipalPolicy.WindowsPrincipal);
            
            // All threads should now have a Windows principal
            // according to the documentation on https://docs.microsoft.com/en-us/dotnet/api/system.appdomain.setprincipalpolicy?view=netcore-3.1
            // Works here.
            var t = new Thread(PrintPrincipalInformation);
            t.Start();
            t.Join();

            //But doesn't work here anymore
            t = new Thread(PrintPrincipalInformation);
            t.Start();
            t.Join();

            // Create a principal to use for new threads.
            IIdentity identity = new GenericIdentity("NewUser");
            IPrincipal principal = new GenericPrincipal(identity, null);
            AppDomain.CurrentDomain.SetThreadPrincipal(principal);
            
            // Should create a new thread with the principal created above, 
            // according to the documentation here: https://docs.microsoft.com/en-us/dotnet/api/system.appdomain.setthreadprincipal?view=netcore-3.1
            // Also doesn't work.
            t = new Thread(PrintPrincipalInformation);
            t.Start();
            t.Join();
        
            // Wait for user input before terminating.
            Console.ReadLine();
        }

        static void PrintPrincipalInformation()
        {
            var curPrincipal = Thread.CurrentPrincipal;
            if(curPrincipal != null)
            {
                Console.WriteLine("Type: " + curPrincipal.GetType().Name);
                Console.WriteLine("Name: " + curPrincipal.Identity.Name);
                Console.WriteLine("Authenticated: " +
                                  curPrincipal.Identity.IsAuthenticated);
                Console.WriteLine();
                return;
            }
            Console.WriteLine("Thread.CurrentPrincipal is NULL");
        }
    }
```

.NET Core Runtime 3.1.1
.NET Core SDK 3.1.101

**Comment by jkotas** [comment]
Regression introduced by https://github.com/dotnet/corefx/pull/34747 here:
https://github.com/dotnet/corefx/pull/34747/files#diff-e388668f95442712879e252fb0d48001R193

cc @MarcoRossignoli

**Comment by MarcoRossignoli** [comment]
I'll take a look today, thank's.

**Comment by jkotas** [comment]
@marcselis @gomathip02 Do you have a workaround for this bug, or is this blocking you from using .NET Core 3.1 for your application?

**Comment by marcselis** [comment]
> @marcselis @gomathip02 Do you have a workaround for this bug, or is this blocking you from using .NET Core 3.1 for your application?

Sorry for the late reply, I am on vacation.
No workaround yet.   For now we can live without the identity in the current thread because we are still developing.  In a few weeks we will be going to production and then we'll need it.  I saw the big is already fixed. I hope a new 3.1 version will be released soon...

**Comment by jkotas** [comment]
This bug is fixed for .NET 5. It is not fixed for .NET Core 3.1. .NET Core 3.1 is LTS and so it is only getting bug fixes that have a very strong justification. If you would like to see the fix ported to .NET Core 3.1, would you mind writing a small paragraph why this fix is critical for you and why it is not feasible to work around it for your app?

**Comment by marcselis** [comment]
Ok, I'll write it in the beginning of next week when I'm back at work.

**Comment by mgnsm** [comment]
I believe it's "a very strong justification" to fix this in .NET Core 3.1 as it's currently not possible to implement authorization in a desktop app by setting the application thread's default identity as explained in [this blog post](https://blog.magnusmontin.net/2013/03/24/custom-authorization-in-wpf/). 

This effectivly makes the `Thread.CurrentPrincipal` API pretty useless in desktop apps that targets .NET Core 3.1 and stops people from migrating from the .NET Framework.

**Comment by marcselis** [comment]
We have 100's of applications that all use a common framework to talk to databases and queues.  This common framework is compiled against .NET Standard 2.0 and uses Thread.CurrentPrincipal for authentication and for audit logging. Most of its code is async.

This common framework works ok in .NET Framework 4.6.1 to 4.8 applications.
But this bug is preventing us from migrating any of those existing applications to a LTS support .NET Core version or to develop any new application in .NET Core.

**Comment by eiriktsarpalis** [comment]
@marcselis we have backported the fix to the 3.1 branch and will be released soon in a servicing update. I think we can close this issue now?

**Comment by marcselis** [comment]
Ok, thanks for the backport!
